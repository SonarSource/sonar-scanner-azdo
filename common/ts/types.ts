export enum EndpointType {
  SonarCloud = 'SonarCloud',
  SonarQube = 'SonarQube'
}
