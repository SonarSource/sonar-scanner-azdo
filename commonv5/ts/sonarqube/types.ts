export type ProjectStatus = {
  status: string;
  conditions: Condition[];
};

export type AnalysisResult = {
  dashboardUrl?: string;
  projectName?: string;
  metrics?: Metric[];
  warnings: string[];
};

export type Condition = {
  status: string;
  metricKey: string;
  actualValue?: string;
  comparator?: string;
  periodIndex?: number;
  errorThreshold?: string;
  warningThreshold?: string;
};

export type Metric = {
  custom?: boolean;
  decimalScale?: number;
  description?: string;
  direction?: number;
  domain?: string;
  hidden?: boolean;
  key: string;
  name: string;
  qualitative?: boolean;
  type: string;
};

export type MetricsResponse = {
  metrics: Metric[];
  p: number;
  ps: number;
  total: number;
};
