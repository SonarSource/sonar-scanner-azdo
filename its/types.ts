export type YamlContent = { [key: string]: unknown };

export type TaskDefinition = Partial<{
  task: string;
  script: string;
  workingDirectory: string;
  inputs: YamlContent;
}>;

export type PipelineCombination = {
  // OS to run on
  os: "unix" | "windows";

  // Version of the tasks to use (incl. extension)
  version:
    | {
        extension: "sonarcloud";
        version: 1 | 2 | 3;
      }
    | {
        extension: "sonarqube" | "sonarqube:lts";
        version: 5 | 6 | 7;
      };

  // Scanner to use
  scanner:
    | {
        type: "cli" | "dotnet";
        version?: string;
      }
    | {
        type: "other";
        subtype: "gradle" | "maven";
      };
};
