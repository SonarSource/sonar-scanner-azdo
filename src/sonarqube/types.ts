export type ProjectStatus = {
  status: string;
  conditions: Condition[];
};

export type AnalysisResult = {
  dashboardUrl?: string;
  projectName?: string;
  metrics?: Metric[];
  warnings: string[];
};

export type Condition = {
  status: string;
  metricKey: string;
  actualValue?: string;
  comparator?: string;
  periodIndex?: number;
  errorThreshold?: string;
  warningThreshold?: string;
};

export type Metric = {
  custom?: boolean;
  decimalScale?: number;
  description?: string;
  direction?: number;
  domain?: string;
  hidden?: boolean;
  key: string;
  name: string;
  qualitative?: boolean;
  type: string;
};

export type MetricsResponse = {
  metrics: Metric[];
  p: number;
  ps: number;
  total: number;
};

export interface PeriodMeasure {
  bestValue?: boolean;
  index: number;
  value: string;
}

export interface Measure {
  metric: string;
  bestValue?: boolean;
  period?: PeriodMeasure;
  value?: string;
}

export interface MeasureResponse {
  component: {
    measures: Measure[];
  };
  [key: string]: unknown;
}
