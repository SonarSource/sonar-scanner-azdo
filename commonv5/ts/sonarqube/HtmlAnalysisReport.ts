import * as tl from "azure-pipelines-task-lib/task";
import {
  htmlDiv,
  htmlLink,
  htmlMainDiv,
  htmlMetricList,
  htmlMetricListItem,
  htmlQualityGateHeader,
  htmlSectionDiv,
  htmlSeparator,
} from "../helpers/html";
import { formatMeasure } from "../helpers/measures";
import { AnalysisResult, ProjectStatus } from "./types";

export default class HtmlAnalysisReport {
  private readonly projectStatus: ProjectStatus;
  private readonly result: AnalysisResult;

  public static getInstance(
    projectStatus: ProjectStatus,
    result: AnalysisResult,
  ): HtmlAnalysisReport {
    return new HtmlAnalysisReport(projectStatus, result);
  }

  constructor(projectStatus: ProjectStatus, result: AnalysisResult) {
    this.projectStatus = projectStatus;
    this.result = result;
  }

  public getFailedConditions() {
    return this.projectStatus.conditions.filter((condition) =>
      ["WARN", "ERROR"].includes(condition.status.toUpperCase()),
    );
  }

  public getHtmlAnalysisReport() {
    tl.debug("Generate analysis report.");

    const html = htmlMainDiv(
      [
        htmlQualityGateHeader(this.projectStatus, this.result.projectName),
        htmlSeparator(),
        this.getHtmlQualityGateDetailSection(),
        this.getHtmlDashboardLink(),
      ].join(""),
    );

    return html;
  }

  private getHtmlQualityGateDetailFailedSection() {
    const rows = this.getFailedConditions().map((condition) => {
      const metric = this.result.metrics.find((metric) => metric.key === condition.metricKey);
      if (!metric) {
        return null;
      }
      const threshold =
        condition.status === "WARN" ? condition.warningThreshold : condition.errorThreshold;
      const requiredContent =
        (metric.type !== "RATING" ? formatMeasure(condition.comparator, "COMPARATOR") + " " : "") +
        formatMeasure(threshold, metric.type);

      return htmlMetricListItem(
        "❌",
        `${formatMeasure(condition.actualValue, metric.type)} ${metric.name}`,
        requiredContent,
      );
    });

    return rows.length > 0 ? htmlSectionDiv("Failed Conditions", htmlMetricList(rows)) : "";
  }

  private getHtmlMeasureListItem(icon: string, metricKey: string) {
    const condition = this.projectStatus.conditions.find((c) => c.metricKey === metricKey);
    const metric = this.result.metrics.find((m) => m.key === metricKey);
    if (!condition || !metric) {
      return "";
    }

    return htmlMetricListItem(
      icon,
      `${formatMeasure(condition.actualValue, metric.type)} ${metric.name}`,
    );
  }

  private getHtmlQualityGateDetailPassedSection() {
    const issuesItems = [
      this.getHtmlMeasureListItem("✅", "new_violations"),
      this.getHtmlMeasureListItem("🔧", "pullrequest_addressed_issues"),
      this.getHtmlMeasureListItem("💤", "new_accepted_issues"),
    ].filter((item) => item.length > 0);
    const issuesSection =
      issuesItems.length > 0 ? htmlSectionDiv("Issues", htmlMetricList(issuesItems)) : "";

    const measuresItems = [
      this.getHtmlMeasureListItem("✅", "new_security_hotspots"),
      this.getHtmlMeasureListItem("✅", "new_coverage"),
      this.getHtmlMeasureListItem("✅", "new_duplicated_lines_density"),
    ].filter((item) => item.length > 0);
    const measuresSection =
      measuresItems.length > 0 ? htmlSectionDiv("Measures", htmlMetricList(measuresItems)) : "";

    return issuesSection + measuresSection;
  }

  private getHtmlQualityGateDetailSection() {
    if (!this.result.metrics) {
      return "";
    }

    if (["WARN", "ERROR"].includes(this.projectStatus.status)) {
      return this.getHtmlQualityGateDetailFailedSection();
    } else {
      return this.getHtmlQualityGateDetailPassedSection();
    }
  }

  private getHtmlDashboardLink() {
    if (!this.result.dashboardUrl) {
      return "";
    }

    return htmlDiv(
      `margin-top: 24px;
      padding-left: 28px;`,
      htmlLink("", this.result.dashboardUrl, "See analysis details on SonarQube"),
    );
  }
}
