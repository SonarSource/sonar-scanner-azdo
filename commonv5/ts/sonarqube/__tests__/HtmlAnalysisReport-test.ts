import HtmlAnalysisReport from "../HtmlAnalysisReport";
import { AnalysisResult, Metric } from "../types";

const MOCKED_METRICS: Metric[] = [{ key: "new_violations", name: "Bugs", type: "INT" }];
const MOCKED_PROJECT_STATUS_ERROR = {
  status: "ERROR",
  conditions: [
    {
      status: "ERROR",
      metricKey: "new_violations",
      comparator: "GT",
      errorThreshold: "0",
      actualValue: "1",
    },
  ],
};
const MOCKED_PROJECT_STATUS_SUCCESS = {
  status: "OK",
  conditions: [
    {
      status: "OK",
      metricKey: "new_violations",
      comparator: "GT",
      errorThreshold: "0",
      actualValue: "0",
    },
  ],
};
const MOCKED_ANALYSIS_RESULT: AnalysisResult = {
  dashboardUrl: "https://dashboard.url",
  projectName: null,
  metrics: MOCKED_METRICS,
  warnings: [],
};

jest.mock("azure-pipelines-task-lib/task", () => ({
  debug: jest.fn(),
  error: jest.fn(),
  getHttpProxyConfiguration: jest.fn().mockImplementation(() => null),
}));

it("should generate an analysis status with error", () => {
  const analysis = new HtmlAnalysisReport(MOCKED_PROJECT_STATUS_ERROR, [], MOCKED_ANALYSIS_RESULT);

  expect(analysis.getFailedConditions()).toHaveLength(1);
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});

it("should not fail when metrics are missing", () => {
  const analysis = new HtmlAnalysisReport(MOCKED_PROJECT_STATUS_ERROR, [], {
    ...MOCKED_ANALYSIS_RESULT,
    dashboardUrl: undefined,
    metrics: undefined,
  });

  expect(analysis.getFailedConditions()).toHaveLength(1);
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});

it("should render passing quality gate measures correctly", () => {
  const analysis = new HtmlAnalysisReport(
    MOCKED_PROJECT_STATUS_SUCCESS,
    [
      {
        metric: "new_violations",
        value: "1",
      },
      {
        metric: "pull_request_fixed_issues",
        period: {
          value: "2",
          index: 1,
        },
      },
      {
        metric: "new_coverage",
        value: "23",
      },
    ],
    MOCKED_ANALYSIS_RESULT,
  );

  expect(analysis.getFailedConditions()).toHaveLength(0);
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});

it("should display the project name", () => {
  const analysis = new HtmlAnalysisReport(MOCKED_PROJECT_STATUS_ERROR, [], {
    ...MOCKED_ANALYSIS_RESULT,
    projectName: "project_name",
  });

  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});
