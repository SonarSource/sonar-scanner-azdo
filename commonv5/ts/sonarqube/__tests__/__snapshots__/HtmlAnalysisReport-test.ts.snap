// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the project name 1`] = `
"<div style="font-family: Inter;
    font-size: 14px;
    font-style: normal;
    line-height: normal;
    font-weight: 400;
    color: rgb(29, 33, 47);
    padding-bottom: 1px;"><div style="font-size: 21px;
    font-weight: 600;"><span>❌</span><span style="
        height: 100%;
        padding-top: 4px;
        margin-left: 4px;
      ">Quality Gate failed (project_name)</span></div><div style="background: #EBECF0;
    height: 1px;
    margin: 16px 0px;"></div><div style="margin-top: 24px;"><div style="margin-bottom: 12px;">Failed Conditions</div><ul
    style="
      padding: 0;
      margin: 0;
    "
  ><li
      style="
        list-style: none;
        align-items: center;
        margin-top: 8px;
      "
    ><span style="width: 24px">❌</span>
<span style="margin-left: 4px">1 Bugs</span>
<span class="margin-left: 12px">(required &le; 0)</span></li></ul></div><div style="margin-top: 24px;
      padding-left: 28px;"><a
    style="
      color: rgb(9, 105, 218);
      text-decoration: none;
      border-bottom: 1px solid rgb(9, 105, 218);
      font-weight: 500;
      width: fit-content; 
    "
    href="https://dashboard.url"
  >See analysis details on SonarQube</a></div></div>"
`;

exports[`should generate an analysis status with error 1`] = `
"<div style="font-family: Inter;
    font-size: 14px;
    font-style: normal;
    line-height: normal;
    font-weight: 400;
    color: rgb(29, 33, 47);
    padding-bottom: 1px;"><div style="font-size: 21px;
    font-weight: 600;"><span>❌</span><span style="
        height: 100%;
        padding-top: 4px;
        margin-left: 4px;
      ">Quality Gate failed</span></div><div style="background: #EBECF0;
    height: 1px;
    margin: 16px 0px;"></div><div style="margin-top: 24px;"><div style="margin-bottom: 12px;">Failed Conditions</div><ul
    style="
      padding: 0;
      margin: 0;
    "
  ><li
      style="
        list-style: none;
        align-items: center;
        margin-top: 8px;
      "
    ><span style="width: 24px">❌</span>
<span style="margin-left: 4px">1 Bugs</span>
<span class="margin-left: 12px">(required &le; 0)</span></li></ul></div><div style="margin-top: 24px;
      padding-left: 28px;"><a
    style="
      color: rgb(9, 105, 218);
      text-decoration: none;
      border-bottom: 1px solid rgb(9, 105, 218);
      font-weight: 500;
      width: fit-content; 
    "
    href="https://dashboard.url"
  >See analysis details on SonarQube</a></div></div>"
`;

exports[`should not fail when metrics are missing 1`] = `
"<div style="font-family: Inter;
    font-size: 14px;
    font-style: normal;
    line-height: normal;
    font-weight: 400;
    color: rgb(29, 33, 47);
    padding-bottom: 1px;"><div style="font-size: 21px;
    font-weight: 600;"><span>❌</span><span style="
        height: 100%;
        padding-top: 4px;
        margin-left: 4px;
      ">Quality Gate failed</span></div><div style="background: #EBECF0;
    height: 1px;
    margin: 16px 0px;"></div></div>"
`;

exports[`should render passing quality gate measures correctly 1`] = `
"<div style="font-family: Inter;
    font-size: 14px;
    font-style: normal;
    line-height: normal;
    font-weight: 400;
    color: rgb(29, 33, 47);
    padding-bottom: 1px;"><div style="font-size: 21px;
    font-weight: 600;"><span>✅</span><span style="
        height: 100%;
        padding-top: 4px;
        margin-left: 4px;
      ">Quality Gate passed</span></div><div style="background: #EBECF0;
    height: 1px;
    margin: 16px 0px;"></div><div style="margin-top: 24px;"><div style="margin-bottom: 12px;">Issues</div><ul
    style="
      padding: 0;
      margin: 0;
    "
  ><li
      style="
        list-style: none;
        align-items: center;
        margin-top: 8px;
      "
    ><span style="width: 24px">✅</span>
<span style="margin-left: 4px">1 new issues</span>
</li></ul></div><div style="margin-top: 24px;
      padding-left: 28px;"><a
    style="
      color: rgb(9, 105, 218);
      text-decoration: none;
      border-bottom: 1px solid rgb(9, 105, 218);
      font-weight: 500;
      width: fit-content; 
    "
    href="https://dashboard.url"
  >See analysis details on SonarQube</a></div></div>"
`;
